src/2decompfft/decomp.f90

```diff
@@ -0,0 +1,113 @@
+submodule(m_decomp) m_decomp_2decompfft
+!! Parallel decomposition provided by 2decomp&FFT
+
+  use mpi
+  implicit none
+
+  type, extends(decomp_t) :: decomp_2decompfft_t
+  contains
+    procedure :: decomposition => decomposition_2decompfft
+  end type
+
+  contains
+
+  module subroutine init_decomp(decomp)
+    class(decomp_t), allocatable, intent(out):: decomp
+
+    allocate(decomp_2decompfft_t :: decomp)
+  end subroutine
+
+  module subroutine decomposition_2decompfft(self, grid, par)
+    !! Performs 2D mesh decomposition using 2decomp&fft
+    use m_mesh_content, only: par_t, grid_t
+    use decomp_2d, only: decomp_2d_init, DECOMP_2D_COMM_CART_X, xsize, xstart
+    use decomp_2d_mpi, only: nrank, nproc
+
+    class(decomp_2decompfft_t) :: self
+    class(grid_t), intent(inout) :: grid
+    class(par_t), intent(inout) :: par 
+    integer :: p_col, p_row
+    integer, allocatable, dimension(:, :, :) :: global_ranks
+    integer, allocatable, dimension(:) :: global_ranks_lin
+    integer :: nproc
+    integer, dimension(3) :: subd_pos, subd_pos_prev, subd_pos_next
+    logical, dimension(3) :: periodic_bc
+    integer :: dir
+    logical :: is_last_domain
+    integer :: nx, ny, nz
+    integer :: ierr
+    integer :: cart_rank
+    integer, dimension(2) :: coords
+
+    if (par%is_root()) then
+      print*, "Domain decomposition by 2decomp&fft"
+    end if
+    nrank = par%nrank
+    nproc = par%nproc
+
+    nx = grid%global_cell_dims(1)
+    ny = grid%global_cell_dims(2)
+    nz = grid%global_cell_dims(3)
+
+    p_row = par%nproc_dir(2)
+    p_col = par%nproc_dir(3)
+    if (p_row*p_col /= par%nproc) then
+      error stop "Decomposition in X not supported by 2decomp&fft backend"
+    end if
+    periodic_bc(:) = grid%periodic_BC(:)
+    call decomp_2d_init(nx, ny, nz, p_row, p_col, periodic_bc)
+
+    ! Get global_ranks
+    allocate(global_ranks(1, p_row, p_col))
+    allocate(global_ranks_lin(p_row*p_col))
+    global_ranks_lin(:) = 0
+
+    call MPI_Comm_rank(DECOMP_2D_COMM_CART_X, cart_rank, ierr)
+    call MPI_Cart_coords(DECOMP_2D_COMM_CART_X, cart_rank, 2, coords, ierr)
+
+    global_ranks_lin(coords(1)+1 + p_row*(coords(2))) = par%nrank
+
+    call MPI_Allreduce(MPI_IN_PLACE, global_ranks_lin, p_row*p_col, MPI_INTEGER, MPI_SUM, MPI_COMM_WORLD, ierr)
+
+    global_ranks = reshape(global_ranks_lin, shape=[1, p_row, p_col])
```

Comment:
I think the main issue is that 2decomp doesn't only gives us the rank mapping, it also sets the number of cells in each direction per rank and the cell offset. These two things are more mesh related which makes this function a bit more tied in with the mesh object and initialisation. I am still thinking about the best way to do this and in particular reducing the code duplication between the generic and 2decomp decompositions.

Reply:
I think we can pass cell numbers as well as rank mapping when instantantiating the mesh object. The offset can then be determined easily from the cell numbers during the mesh object initialisation. Is there anything else we ask 2decomp to provide?

What I have in mind is initialising 2decomp somewhere in the main program just before mesh. Then we have the outputs from 2decomp: rank mapping and local cell sizes. We can pass these two as arguments when instantiating the mesh object. We can have these as optional arguments so we fall back to default trivial mappings in tests. Also, we can do this 2decomp initialision with ifdefs, so it is called only when x3d2 is compiled with 2decomp support. If mesh object is not given optional domain mapping info and cell count info we use our current strategy. (or when domain mapping and cell data is all zeros, meaning no preference just to make it easier)

Reply:
yeah, sure. That's doable. I will try to avoid ifdef though, the current implementation was designed to not have any ifdef and only rely on the linking to select which implementation to choose from. I feel like that's cleaner and allows us to add more decomposition methods later on without any change to the current code.
This means that, even without 2decomp, the generic decomposition will still be done prior to the mesh object instantiation etc.

Overall, I am not sure I fully see the benefits of having the decomposition outside the mesh init as the two are intertwined to me, but that's fine, I will do that.

Reply:
I think the benefit is that we can remove `decomp.f90`, `2decompfft/decomp.f90`, and `decomp_generic.f90` files, and this new structure around `m_decomp` modules if we allow mesh to take a 'suggestion' regarding decomposition. We can still have any custom decomposition and can implement this inside the mesh.

Not having ifdef is better definitely, can you comment on whether this would allow us to change between FFT and iterative solver, and potentially changing between using 2decomp or a different library (more related to CUDA backend, but better to not prevent this with the structure we design here).

Reply:
I mean changing between FFT/iteretive solver at runtime. For example its likekly we'll have a case with stretching in two/three directions, and then FFT wouldn't be compatible with that case, and we'll have to use iterative with it. Ideally we shouldn't make a compile time decision between these solvers and should be select at runtime.

Reply:
We might want to have a discussion about how to handle two different Poisson solvers (at compilation or at runtime or something else), in terms of user friendliness, performance, and ease of implementation :-)

Reply:
Yeah, having a chat about it is probably a good idea, also because @pbartholomew08 has implemented something similar for his iterative poisson solver. It would be good to coordinate to have a consistent solution for these things.

Reply:
I can build my branch without PETSc, it requires a dummy module to link against which will simply crash the program if you try to use iterative solvers (untested). And it should be possible to switch at runtime between FFT/CG (untested).

However, this may be relatively easier to do than with a low level data structure as in x3d2-mesh vs 2decomp - perhaps if we want that kind of flexibility we need to accept data copies between "x3d2 space" and "2decomp space", this is currently what I have in the PETSc module too, but I hope to eliminate.

The dummy module is simply to enable compilation:
```
submodule(m_poisson_cg) m_dummy_poisson_cg
  use m_base_backend, only: base_backend_t
contains
  module subroutine init_precon_impl(precon, backend)
    class(poisson_precon_impl_t), allocatable, intent(out) :: precon
    class(base_backend_t), intent(in) :: backend
    error stop "This build doesn't support iterative Poisson solvers"
  end subroutine
  module subroutine apply_precon_internal(self, p, b, backend)
    class(poisson_precon_impl_t) :: self
    class(field_t), intent(in) :: p ! Pressure solution
    class(field_t), intent(inout) :: b ! The evaluated matrix-vector product
    class(base_backend_t), intent(in) :: backend
    error stop "This build doesn't support iterative Poisson solvers"
  end subroutine apply_precon_internal
  module subroutine init_solver(solver, backend)
    class(poisson_solver_t), allocatable, intent(out) :: solver
    class(base_backend_t), target, intent(in) :: backend
    error stop "This build doesn't support iterative Poisson solvers"
  end subroutine init_solver
  module subroutine solve_poisson(self, p, f, backend)
    class(poisson_solver_t) :: self
    class(field_t), intent(inout) :: p ! Pressure solution
    class(field_t), intent(in) :: f    ! Poisson RHS
    class(base_backend_t), intent(in) :: backend
    error stop "This build doesn't support iterative Poisson solvers"
  end subroutine solve_poisson
end submodule
```

Reply:
I think the main difference currently between my implementation and Paul's one is that I don't have a dummy module, instead, I have the 'generic' decomposition implementation that is being used when 2decomp isn't available. This doesn't allow runtime switch.

So what I am going to do now is have that dummy module like Paul and move the generic decomposition somewhere else that is always compiled. That way we can choose at run time which one we use while still avoiding ifdef etc.
For now I will leave this inside the mesh initialisation, but it can easily be moved outside of it later on if needed.

Reply:
It's been decided that to stick with a submodule method for now and we might move to a different method in a future PR.